streamer/m3u8_concater.py

```diff
@@ -0,0 +1,955 @@
+    
+    # We will fill this map, file names to (OutputStream)s and will give it out
+    # to each media playlist we initialize.  Each media playlist then will pick
+    # the right output stream for itself.
+    streams_map: Dict[str, OutputStream] = {}
+    
+    # Add a mapping between single segment file names and their output stream.
+    streams_map.update({
+        outstream.get_single_seg_file().write_end(): outstream
+        for outstream in packager.output_streams
+      })
+    # Add another mapping between the first segment in multi-segment file names
+    # and their corresponding output streams.
+    streams_map.update({
+        outstream.get_media_seg_file()
+        .write_end()
+        .replace('$Number$', '1'): outstream
+        for outstream in packager.output_streams
+      })
+    
+    dir_name = os.path.dirname(file_name)
+    
+    with open(file_name, 'r') as master_playlist:
+      line = master_playlist.readline()
+      while line:
+        if line.startswith('#EXT-X-MEDIA'):
+          stream_info = _extract_attributes(line)
+          self.playlists.append(MediaPlaylist(stream_info, dir_name,
+                                              output_dir,
+                                              streams_map))
+        elif line.startswith('#EXT-X-STREAM-INF'):
+          stream_info = _extract_attributes(line)
+          # Quote the URI to keep consistent,
+          # as the URIs in EXT-X-MEDIA are quoted too.
+          stream_info['URI'] = _quote(master_playlist.readline().strip())
+          self.playlists.append(MediaPlaylist(stream_info, dir_name,
+                                              output_dir,
+                                              streams_map))
+        line = master_playlist.readline()
+      # Get the master playlist duration from an arbitrary stream.
+      self.duration = self.playlists[-1].duration
+  
+  def write(self, file: str, comment: str) -> None:
+    """Writes the master playlist and the nested media playlists in
+    the file system.
+    """
+    
+    dir_name = os.path.dirname(file)
+    with open(file, 'w') as master_playlist:
+      content = MasterPlaylist.header
+      content += comment
+      # Write #EXT-X-MEDIA media playlists first.
+      for media_playlist in self.playlists:
+        if media_playlist.stream_info.get('TYPE'):
+          media_playlist.write(dir_name)
+          content += '#EXT-X-MEDIA:' + ','.join(
+              [key + '=' + value for
+               key, value in media_playlist.stream_info.items()]) + '\n'
+      content += '\n'
+      # Then write #EXT-X-STREAM-INF media playlists.
+      for media_playlist in self.playlists:
+        if media_playlist.stream_info.get('TYPE') is None:
+          media_playlist.write(dir_name)
+          # We don't write the URI in the attributes of a stream
+          # variant playlist.  Pop out the URI.
+          uri = _unquote(media_playlist.stream_info.pop('URI'))
```

Comment:
This is really cool.  I had no idea you could "pop" from a dictionary in Python.